### PR TITLE
Autograd deadlock for recent glibc fix

### PR DESCRIPTION
**Pull request by albanD**
* Remove compilation warning in `csrc/Module.cpp`
* Remove compilation warnings in `lib/libshm/manager.cpp`
* Make the `ReadyQueue` passed to worker threads in the autograd engine `shared_ptr` so that they get destroyed properly (without causing any deadlocks like #1233)

### CODE REVIEW

**Comment by soumith** [comment]
super stuff Alban!

File: torch/lib/libshm/manager.cpp [hunk 1]
```diff
@@ -55,8 +55,9 @@ void unregister_fd(int fd) {
 
 
 void print_init_message(const char *message) {
-  write(1, message, strlen(message));
-  write(1, "\n", 1);
+  size_t unused;
+  unused = write(1, message, strlen(message));
```
Comment (apaszke) [inline torch/lib/libshm/manager.cpp]: You can just cast the call to void

File: torch/lib/libshm/manager.cpp [hunk 2]
```diff
@@ -109,7 +110,7 @@ int main(int argc, char *argv[]) {
     if (nevents == 0 && client_sessions.size() == 0)
       break;
 
-    for (struct pollfd &pfd: pollfds) {
+    for (auto &pfd: pollfds) {
```
Comment (apaszke) [inline torch/lib/libshm/manager.cpp]: Was it incorrect?

**Review by apaszke — COMMENTED** [review]
The problem with current implementation is that it will leak the threads and ready queues, but I guess it's ok for now as we only have a single global queue. Can you only leave a comment in Engine's destructor?

File: torch/lib/libshm/manager.cpp (hunk 2 above)
Comment (albanD) [inline torch/lib/libshm/manager.cpp]: It was correct bug giving this warning:
```
/pytorch/torch/lib/libshm/manager.cpp: In function 'int main(int, char**)':
/pytorch/torch/lib/libshm/manager.cpp:112:10: warning: types may not be defined in a for-range-declaration
     for (struct pollfd &pfd: pollfds) {
          ^~~~~~
```

File: torch/lib/libshm/manager.cpp (hunk 1 above)
Comment (albanD) [inline torch/lib/libshm/manager.cpp]: This is not supported anymore by gcc (well I can do it but the warning will still be there)
And it looks like a quite controversial subject https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425

**Comment by albanD** [comment]
@apaszke I think its ok to leak threads on destruction as the Engine should never be destroyed anyway (the thread creation logic would fail if we destroy an engine and create another one).
I added a comment for this.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Awesome! Thanks Alban!

**Comment by tylergenter** [comment]
I'm now getting a really weird deadlock in autograd on Windows. I wasn't having this problem a week ago, and I'm thinking this patch is the problem (although I'm not positive). Basically, a worker thread is throwing a runtime error, and everything goes to hell while that error is being propagated. Commenting out python_error::~python_error fixes the problem. What I think is happening is that the python_error destructor is being called before THPEngine_run_backward can process the error. For a test case, line 816 in test_nn.py deliberately causes the assert in THPFunction_saved_tensors to fail. I don't understand this code well enough to come up with a solution, or even be sure that this is the problem.

**Comment by apaszke** [comment]
Do you know where do the threads stop? It's probably because `python_error`'s destructor acquires the GIL, but I can't really think of any way to make it deadlock right now. Maybe we acquire `backward_task.mutex` and the GIL in different orders.

**Comment by tylergenter** [comment]
The main thread returns to the Python interpreter, but it acts really weird. Basically I have to kill the process. The worker thread waits in ReadyQueue::pop_back, looks like it's waiting for the queue to not be empty.

What is happening is that when Engine::thread_on_exception gets called in the worker thread, the exception in python_error looks normal. More specifically, e->type->ob_refcount == 31, and e->value->ob_refcount == e->traceback->ob_refcount = 1. When we get to Engine::backward in the main thread (line 339), e->type->ob_refcount == 30, and e->value->ob_refcount == e->traceback->ob_refcount = 0. By the time we get to python_error::restore (via THPEngine_run_backward), e->type->ob_refcount == 29, and e->value->ob_refcount == e->traceback->ob_refcount = -1. So it appears as if the python_error destructor is getting called twice before python_error::restore, when it shouldn't be called at all.

**Comment by apaszke** [comment]
I guess that it's because python_error is copyable, but the copy constructor doesn't INCREF the held objects. We should either make it only movable or add a custom copy constructor.

**Comment by tylergenter** [comment]
Adding a copy constructor fixed my problem.

**Pull request merged**